server/eclipse-project/src/main/java/us/freeandfair/corla/auth/AbstractAuthentication.java

```diff
@@ -0,0 +1,237 @@
+                                            AuthenticationInterface.PASSWORD);
+            Main.LOGGER.info("Second factor authentication failed for administrator" + 
+                the_username);
+            result = false;
+          }
+        }
+      }
+    } catch (final PersistenceException e) {
+      // there's nothing we can really do here other than saying that the
+      // authentication failed; it's also possible we failed to update the last
+      // login time, but that's not critical
+      deauthenticate(the_request, the_username);
+    }
+
+    if (!result) {
+      // a failed authentication attempt removes any existing session authentication 
+      deauthenticate(the_request, the_username);
+      Main.LOGGER.info("Authentication failed for user " + the_username);
+    }
+
+    return result;
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public County authenticatedCounty(final Request the_request) {
+    County result = null;
+    if (isAuthenticatedAs(the_request, AdministratorType.COUNTY,
+                    the_request.queryParams(AuthenticationInterface.USERNAME))) {
+      final Administrator admin = 
+          (Administrator) the_request.session().attribute(ADMIN);
+      if (admin != null) {
+        result = CountyQueries.forAdministrator(admin);
+      }
+    }
+    return result;
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void traditionalDeauthenticate(final Request the_request,
+                                        final String the_username) {
+    the_request.session().removeAttribute(ADMIN);
+    Main.LOGGER.info("session is now traditionally deauthenticated");
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void twoFactorDeauthenticate(final Request the_request,
+                                      final String the_username) {
+    the_request.session().removeAttribute(ADMIN);
+    Main.LOGGER.info("session is now second factor deauthenticated");
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public boolean traditionalAuthenticated(final Request the_request,
+                                          final String the_username) {
+    final String auth_stage = 
+        the_request.session().attribute(AuthenticationInterface.AUTH_STAGE);
+    return auth_stage != null && 
+        (auth_stage.equals(AuthenticationInterface.SECOND_FACTOR) ||
+            auth_stage.equals(AuthenticationInterface.ADMIN));
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public boolean secondFactorAuthenticated(final Request the_request,
+                                           final String the_username) {
+    final String auth_stage = 
+        the_request.session().attribute(AuthenticationInterface.AUTH_STAGE);
+    return auth_stage != null && 
+        auth_stage.equals(AuthenticationInterface.ADMIN);
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void logoutAdministrator(final Request the_request,
```

Comment:
What is this method for (i.e., why not just use `deauthenticate`)?

Reply:
Indeed, it is a refactoring artifact.  I've removed it in e6c81bb.